### PR TITLE
Various optimization in tracking 

### PR DESCRIPTION
**Pull request by VinInn**
Optimization in various areas, not affecting algorithm behavior.
initialStepPreSplitting made faster w/o affecting physics.
Expect negligible regressions due to numerics...
MTV: 
ttbar 0PU with latest IB: 
http://innocent.home.cern.ch/innocent/RelVal/pu0_TkSpeed91X/plots_ootb/effandfake1.pdf
ttbar PU50 with pre2
http://innocent.home.cern.ch/innocent/RelVal/pu50_fastTk91_2/plots_ootb/effandfake1.pdf

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

File: DataFormats/SiPixelCluster/interface/SiPixelCluster.h [hunk 1]
```diff
@@ -127,11 +127,11 @@ class SiPixelCluster {
   int sizeY() const { return colSpan() +1;}
   
   
-  inline float charge() const {
-    float qm = 0.0;
+  inline int charge() const {
+    int qm = 0;
     int isize = thePixelADC.size();
     for (int i=0; i<isize; ++i)
```
Comment (ianna) [inline DataFormats/SiPixelCluster/interface/SiPixelCluster.h]: @VinInn - it can be simplified:
```
for( auto i :  thePixelADC )
 qm += i;
```

Comment (VinInn) [inline DataFormats/SiPixelCluster/interface/SiPixelCluster.h]: or
return std::accumulate(thePixelADC.begin(),thePixelADC.end(),int(0));

File: RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py [hunk 2]
```diff
@@ -128,9 +128,9 @@
 initialStepChi2EstPreSplitting = RecoTracker.MeasurementDet.Chi2ChargeMeasurementEstimator_cfi.Chi2ChargeMeasurementEstimator.clone(
     ComponentName = cms.string('initialStepChi2EstPreSplitting'),
     nSigma = cms.double(3.0),
-    MaxChi2 = cms.double(30.0),
+    MaxChi2 = cms.double(16.0),
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py]: updates in this file look more than just technical optimization numerics.

Please add some MTV or hiteff plots for vfp30 case.

File: DataFormats/SiPixelCluster/interface/SiPixelCluster.h (hunk 1 above)
Comment (ianna) [inline DataFormats/SiPixelCluster/interface/SiPixelCluster.h]: even better

**Comment by VinInn** [comment]
> On 25 Apr, 2017, at 8:24 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> Please add some MTV or hiteff plots for vfp30 case.
could you please remind us which relval to use?
please note this is "presplitting": none of this track will make it to the final collection,
no hiteff monitoring exists.
btw there is no reliable simulation of vfp30 at best of my knowledge

v.

**Comment by slava77** [comment]
On 4/26/17 12:29 AM, Vincenzo Innocente wrote:
> 
>> On 25 Apr, 2017, at 8:24 PM, Slava Krutelyov <notifications@github.com> wrote:
>> 
>> Please add some MTV or hiteff plots for vfp30 case.
> could you please remind us which relval to use?
> please note this is "presplitting": none of this track will make it to
> the final collection,
> no hiteff monitoring exists.
> btw there is no reliable simulation of vfp30 at best of my knowledge

hiteff plots for a 2016E data relval may be enough.

Thank you.

**Comment by VinInn** [comment]
this is the comparison "out of the box" for run 277069
image

now I am running for era.Run2_2016_HIPM

**Comment by VinInn** [comment]
and here is with vfp30 mitigation

**Comment by slava77** [comment]
On 4/27/17 1:02 AM, Vincenzo Innocente wrote:
> and here is with vfp30 mitigation

Thank you.

File: RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h [hunk 3]
```diff
@@ -85,10 +85,10 @@ class PixelCPEBase : public PixelClusterParameterEstimator
    
    struct ClusterParam
    {
-      ClusterParam(const SiPixelCluster & cl) : theCluster(&cl), loc_trk_pred(0.0,0.0,0.0,0.0),
-      probabilityX_(0.0), probabilityY_(0.0), probabilityQ_(0.0), qBin_(0.0),
-      isOnEdge_(false), hasBadPixels_(false), spansTwoROCs_(false), hasFilledProb_(false) {}
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h]: not everything here now have default values or is default constructed.
probs, qBin_, and spansTwoROCs_ are not initialized.
Is it quantifiably faster?
Better keep initialized.

File: RecoTracker/FinalTrackSelectors/plugins/TrackCollectionMerger.cc [hunk 4]
```diff
@@ -290,6 +291,8 @@ namespace {
 
     assert(producer.selTracks_->size()==pmvas->size());
 
+    // std::cout << "TrackCollectionMerger: sel tracks " << producer.selTracks_->size() << std::endl;
```
Comment (slava77) [inline RecoTracker/FinalTrackSelectors/plugins/TrackCollectionMerger.cc]: could be cleaned up

File: RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py [hunk 5]
```diff
@@ -128,9 +128,9 @@
...
     clusterChargeCut = cms.PSet(refToPSet_ = cms.string('SiStripClusterChargeCutLoose')),
-    pTChargeCutThreshold = cms.double(15.)
+#    pTChargeCutThreshold = cms.double(15.)
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

File: RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py [hunk 6]
```diff
@@ -142,8 +142,8 @@
     alwaysUseInvalidHits = True,
     maxCand = 3,
     estimator = cms.string('initialStepChi2Est'),
-    maxDPhiForLooperReconstruction = cms.double(2.0),
-    maxPtForLooperReconstruction = cms.double(0.7)
+#    maxDPhiForLooperReconstruction = cms.double(2.0),
+#    maxPtForLooperReconstruction = cms.double(0.7)
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/InitialStepPreSplitting_cff.py]: same here

File: RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h (hunk 3 above)
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h]: This is intented! initialization is very (very) expensive (I can show you in the assembly the huge amount of "mov" instruction generated). The quantity not initialized are filled just after creation.  In general in CMSSW we are just shuffling bytes, not doing computation, we need to reduce that.

File: RecoTracker/FinalTrackSelectors/plugins/TrackCollectionMerger.cc (hunk 4 above)
Comment (VinInn) [inline RecoTracker/FinalTrackSelectors/plugins/TrackCollectionMerger.cc]: I have LogDebug. In any case it needs to be recompiled. Assume is LogDebug

**Comment by slava77** [comment]
On 4/29/17 12:22 AM, Vincenzo Innocente wrote:
> *@VinInn* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h
> <https://github.com/cms-sw/cmssw/pull/18457#discussion_r114048043>:
> 
>> @@ -85,10 +85,10 @@ class PixelCPEBase : public PixelClusterParameterEstimator
>     
>     struct ClusterParam
>     {
> -      ClusterParam(const SiPixelCluster & cl) : theCluster(&cl), loc_trk_pred(0.0,0.0,0.0,0.0),
> -      probabilityX_(0.0), probabilityY_(0.0), probabilityQ_(0.0), qBin_(0.0),
> -      isOnEdge_(false), hasBadPixels_(false), spansTwoROCs_(false), hasFilledProb_(false) {}
> 
> This is intented! initialization is very (very) expensive (I can show
> you in the assembly the huge amount of "mov" instruction generated). The
> quantity not initialized are filled just after creation. In general in
> CMSSW we are just shuffling bytes, not doing computation, we need to
> reduce that.

Is it cheaper to do memset(this, 0, sizeof(ClusterParam)) ?

**Comment by VinInn** [comment]
ClusterParam) has a virtual destructor: you do not want to do that...

**Comment by Dr15Jones** [comment]
We've also seen related cases where `memset` was called on a memory location that was then used to hold an object and the gcc compiler optimized away the `memset` because they assumed constructors would properly initialize the data that needed to be initialized.

How about adding comments to the initializer list for the uninitialized member functions saying that they are intended to not be initialized there for optization reasons? That way future maintainers of this class will not forget about those members or unoptimize the code?

**Comment by VinInn** [comment]
yes, can be done. Documentation can always been inproved. (still the best documentation is the code itself)

**Comment by slava77** [comment]
On 4/29/17 6:58 AM, Chris Jones wrote:
> We've also seen related cases where |memset| was called on a memory
> location that was then used to hold an object and the gcc compiler
> optimized away the |memset| because they assumed constructors would
> properly initialize the data that needed to be initialized.

Would it similarly outsmart you and remove memset if the data has no
default constructors?

[about vtable pointer, it's a matter of using the right offset instead
of a plain "this"]

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
```
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PrimaryVertexValidation label='PVValidation'
Exception Message:
MissingParameter: Parameter 'Tpurge' not found.
----- End Fatal Exception -------------------------------------------------
Failure running test_OfflineValidaiton_cfg.py: status 73
status = 18688

---> test testAlignmentOfflineValidation had ERRORS
```

not from this PR

**Comment by makortel** [comment]
The unit test failure is likely caused by #18330. It added that parameter, and I see the unit test first failing in  CMSSW_9_1_X_2017-04-28-1100 where the PR was merged.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18457  	974524a
- changes look OK, primarily technical/numerical speedups
- jenkins tests pass and comparisons with baseline show numerous small changes starting from tracks and propagating to downstream objects, nothing is significant to mention (expected)
- local tests with larger samples confirm the expected behavior. The reduction in CPU time in ttbar PU35 (wf 10224) is about 3% total of the reco time with the tracking modules speedup by about 5% (\*Step\* modules ). Tested in CMSSW_9_1_X_2017-04-28-1100.

This PR looks safe for 91X which is essentially closed.
It more naturally falls in the scope of 92X.

**Comment by VinInn** [comment]
I am only afraid the it may clash with changes in pixel geometry and localreco

**Comment by slava77** [comment]
On 5/4/17 4:58 AM, Vincenzo Innocente wrote:
> I am only afraid the it may clash with changes in pixel geometry and
> localreco

OK
doing nothing (not merging now in 91X) should just advance this PR to
the next development release.

**Comment by VinInn** [comment]
ok, let's merge as soon as the first 9_2_X IB appears...

**Comment by VinInn** [comment]
now that 9_2_X IB is out is this PR candidate for merging or the whole test & validation procedure should start from scratch?

**Comment by davidlange6** [comment]
no? did something get reset?

**Comment by VinInn** [comment]
Not to my knowledge...

**Comment by davidlange6** [comment]
so all is working as expected and nothing needs to start over

**Comment by slava77** [comment]
here, jenkins said ``` You potentially added 1979969 lines to the logs ```
... maybe we should put some color coding?

(the fix is in #18617 )

**Pull request merged**